WalnutExternal.lua

```diff
@@ -12,11 +12,45 @@ LibraryDir["VulkanSDK"] = "%{VULKAN_SDK}/Lib"
 Library = {}
 Library["Vulkan"] = "%{LibraryDir.VulkanSDK}/vulkan-1.lib"
 
+
+function linux_dpkg_check_if_package_is_installed(package)
+   if os.isfile("/usr/bin/dpkg-query") then -- It's Debian or a derivative of it:
+      local result, errorCode = os.outputof("/usr/bin/dpkg-query -Wf'${Status}' " .. package)
+      if errorCode ~= 0 or result ~= "install ok installed" then
```

Comment:
Checking for the negative case on specific text is asking for trouble. In my case, "install ok installedinstall ok installed" is returned on one package and that makes this check fail. How are we to know that different Linux distributions don't vary text output?

Reply:
I just googled and found the fix. The correct way is to limit the output to only installed (not uninstalled records), e.g:
```bash
/usr/bin/dpkg-query -Wf'${db:Status-Status}' <package>
```
Does this command work for you?

As long as the distro is based  on Debian (e.g. also Linux Mint, Ubuntu) then with the fix there should be no different output. For other distros, these checks won't be executed because they don't provide dpkg-query (for debian packages)
Reference: https://askubuntu.com/questions/1330078/limit-output-of-dpkg-query-w-to-installed-packages

Reply:
@bernhardkaindl I modified the line to `if errorCode ~= 0 then` and it worked. I checked a failing case by running the function looking for a pretend package as `linux_dpkg_check_if_package_is_installed("nonsense")` and it failed properly.

Reply:
With the fix applied the two lines for the wpkg-query execution and output string check should be:
```py
      local result, errorCode = os.outputof("/usr/bin/dpkg-query -Wf'${db:Status-Status}' " .. package)
      if errorCode ~= 0 or result ~= "installed" then
```

Reply:
I pasted it in and it came up with an error again. Isn't the error code sufficient?

Reply:
> I pasted it in and it came up with an error again. Isn't the error code sufficient?

No, because the error code can also be `0` when the package was installed, but was uninstalled afterwards. e.g:

```bash
sudo apt install -y atop;sudo apt remove -y atop;dpkg-query -l atop;echo $?
...
rc  atop           2.7.1-1      amd64        Monitor for system resources and process activity
0
```

You also need to change the output check to look for "installed" only and no longer "install ok installed",
and because the output could be "install ok" (if installation is possible, but not installed), the guy stackoverflow suggested another improvement:

Reply:
Like this:
```
      local result, errorCode = os.outputof("/usr/bin/dpkg-query -Wf'${db:Status-Status} ${package}' " .. package)
      if errorCode ~= 0 or result ~= "installed " .. package then
```
I'm not sure if my premake5 syntax here is correct, but the correct output would be `installed <name of the package>`

Reply:
I printed out what that result is giving me now: `installed mesa-vulkan-driversinstalled mesa-vulkan-drivers`.

Isn't it *Premake's* job to determine if libraries are available?

Reply:
Use `string.find()` to look for `installed`?

Reply:
> I printed out what that result is giving me now: `installed mesa-vulkan-driversinstalled mesa-vulkan-drivers`.

Ah, not I think I got it: You seem to have an `i386` and and `amd64` version of that package installed, I guess. this should check it, just out of interest.
```
dpkg -l mesa-vulkan-drivers
```

> Isn't it _Premake's_ job to determine if libraries are available?

Ah, well, you could try `https://premake.github.io/docs/os.findlib/`

This should work on any distro.

Reply:
Yeah, both i386 and AMD64. I'm sure I had a good reason to install the older package.

Reply:
> Use `string.find()` to look for `installed`?

Maybe, but that does not mean that the string `installed` was for the same architecture (amd64) that you try to compile for. This should limit the search to am64 versions of the packages (by appending ":amd64" to the package name):

```bash
local result, errorCode = os.outputof("/usr/bin/dpkg-query -Wf'${db:Status-Status} ${package}' " .. package .. ":amd64")
if errorCode ~= 0 and result ~= "installed " .. package then
```

Reply:
> Yeah, both i386 and AMD64. I'm sure I had a good reason to install the older package.

No problem. Because the only "executable" it installs is a python script which may work for `i386` and `amd64`, it should work fine:
```yml
dpkg -L mesa-vulkan-drivers
...
/usr/bin/mesa-overlay-control.py
```